Sanitize ClassDB locking

RandomShaper opened the pull request:
I've realized this issue, where `ClassDB` was still relying on the global lock for some operations instead of its own read-write lock. Even if in practice, due to everything mostly happening on the main thread, this wasn't an issue, the demons were lurking there to cause race conditions as soon as they were given a chance.

This PR removes the macros that were used in the `.cpp` file and uses instead the explicit syntax, which can be used in the header file just as nicely.

I've set this to target 4.4 as I think it's extermely low-risk. **UPDATE:** See latest comments as my initial valoration isn't valid anymore.

RandomShaper committed:
Sanitize ClassDB locking

akien-mga commented:
Seems to crash:

```
Run ./bin/godot.linuxbsd.editor.x86_64.mono --headless --dump-gdextension-interface --dump-extension-api
terminate called after throwing an instance of 'std::system_error'
  what():  Resource deadlock avoided
Aborted (core dumped)
```

```
Run ./bin/godot.windows.editor.x86_64.exe --version
Error: The operation was canceled.
```

RandomShaper commented:
I've updated the PR. I had failed to realize legit use cases that would incur in recursive locking. I've also fixed wrong usage of the lock in `ClassDB::add_property()`, that could lead to a dangling pointer (pointer to `HashMap` element can become invalid if the lock is not held throughout the code dealing with it).

This doesn't qualify as low-risk anymore, does it? 😅

akien-mga commented:
> This doesn't qualify as low-risk anymore, does it? 😅

Yeah I'd be wary to change ClassDB just before the release candidate phase, which we aim to keep short.

Let's merge this for 4.5-dev1 and consider a cherry-pick.

akien-mga commented:
Might warrant a rebase before merge, just in case we've added new stuff to ClassDB using the previous locking (I don't think we did but better safe than sorry).

akien-mga commented:
Rebased.

monxa commented on core/object/class_db.h:
```diff
@@ -152,7 +152,31 @@ class ClassDB {
+			~Lock();
+		};
+	};
+	inline static Locker locker;
```
Isn't the `inline static Locker locker;` declaration redundant?

Since all relevant members (`RWLock lock` and `thread_state`) are already declared as static or thread_local, having an additional inline `static Locker locker` -  seems unnecessary.

What am I missing? :)
